### ISSUE TITLE
Should we move large test cases into separate files?

### ISSUE THREAD

**Issue by petrcermak** [catapult-project/catapult#1893]
I have recently worked on some algorithmic processing components of Trace Viewer that required quite a lot of tests where the **expected outputs** were **large trees and graphs**:

https://github.com/catapult-project/catapult/blob/master/tracing/tracing/model/global_memory_dump_test.html
https://github.com/catapult-project/catapult/blob/master/tracing/tracing/base/multi_dimensional_view_test.html

In their reviews of my latest patch (https://codereview.chromium.org/1585983004/), @primiano and @ruud-v-a suggested that these test cases could be moved to **separate files**. I searched through our codebase and found 3 similar use cases (2 of which seem to be marked as TODOs for removal, #1219?):

https://github.com/catapult-project/catapult/blob/master/perf_insights/perf_insights/ui/pi_app_main_test.html
https://github.com/catapult-project/catapult/blob/master/perf_insights/perf_insights/ui/reports/task_info_report_test.html
https://github.com/catapult-project/catapult/blob/master/perf_insights/perf_insights/ui/reports/weather_report_test.html

I don't have a very strong opinion about this, but it does feel a little strange to check in large HTML files 90% of which are huge object literals.

@natduca, @dj2: What are your views on this? I assume this is even more relevant for PI, which is all about processing lots of data.

**Comment by ruud-v-a** [comment]
My proposal was not necessarily to move tests into a separate file, but if the test cases were generated algorithmically, to put the generator under source control instead of its output.

**Comment by petrcermak** [comment]
I agree it would make sense to check in the test generator(s), but we still need to have its output stored somewhere as well: The generator uses the actual output of the test to generate the expected output, so you can't use it for performing tests (any actual output should match the expected output it induces). While the generator avoids the need to manually write JSON objects, a human still needs to check that the generated test is correct.

**Comment by primiano** [comment]
Then will you need a test for the generator ?

**Comment by petrcermak** [comment]
No, _the human_ (me in this case) is the test for the generator.

**Comment by ruud-v-a** [comment]
> The generator uses the actual output of the test to generate the expected output

Wait — isn’t that backwards? Or is it just to extract some boilerplate?

> While the generator avoids the need to manually write JSON objects, a human still needs to check that the generated test is correct.

Ideally, the generator is simpler than its output, so instead of reviewing 6k lines of generated json, we can review a small generator.

> Then will you need a test for the generator?

It it is simple enough, we should be able to convince ourselves that it is correct without tests.

**Comment by petrcermak** [comment]
>> The generator uses the actual output of the test to generate the expected output
>
> Wait — isn’t that backwards? Or is it just to extract some boilerplate?

The point is that instead of tediously writing the _expected_ structure, you only check that the _actual_ structure is correct (and then copy and paste it as the expectation). That's all there is to it ("test generator" might not be the right name).

I'm not aware of writing a simple test generator that doesn't do the above (unless I write a second implementation of the class under test, which I doubt could be any simpler than the first one).

**Comment by ruud-v-a** [comment]
Ah, I see. So you verified the 6000 lines of json manually and reviewers would need to do that as well?

It could be a sign that the test is trying to test at the wrong level? If you can factor the thing into two orthogonal components with m states and n states, then instead of testing all n*m states, you only need to test m + n states. (So roughly 6000 lines → 150 lines if m ≈ n.)

**Comment by natduca** [comment]
I'm fine adding a test file if it simplifies the testing and clarity... if its giant then maybe its too big to be a unit test? But, if its maintainably large, then by all means... eg layer_tree_host_impl_test_data is a good one that I really don't think we could test any other way.

Minor note: Get its contents with getAsync ideally, instead of sync xhr tho, there're a few tests that do sync still :/
